src/state_machine/state.cpp

```diff
@@ -24,297 +25,192 @@ namespace hyped {
-  data::Navigation nav_data           = data_.getNavigationData();
-  data::Batteries batteries_data      = data_.getBatteriesData();
-  data::Telemetry telemetry_data      = data_.getTelemetryData();
-  data::Motors motors_data            = data_.getMotorData();
-
-  bool encountered_failure = false;
-
-  if (telemetry_data.emergency_stop_command) {
-    encountered_failure = true;
-    log_.ERR("STM", "STOP command received");
-    telemetry_data.emergency_stop_command = false;
-    data_.setTelemetryData(telemetry_data);
-  } else if (nav_data.module_status == ModuleStatus::kCriticalFailure) {
-    encountered_failure = true;
-    log_.ERR("STM", "Critical failure in navigation");
-  } else if (telemetry_data.module_status == ModuleStatus::kCriticalFailure) {
-    encountered_failure = true;
-    log_.ERR("STM", "Critical failure in telemetry");
-  } else if (motors_data.module_status == ModuleStatus::kCriticalFailure) {
-    encountered_failure = true;
-    log_.ERR("STM", "Critical failure in motors");
-  } else if (embrakes_data.module_status == ModuleStatus::kCriticalFailure) {
-    encountered_failure = true;
-    log_.ERR("STM", "Critical failure in embrakes");
-  } else if (batteries_data.module_status == ModuleStatus::kCriticalFailure) {
-    encountered_failure = true;
-    log_.ERR("STM", "Critical failure in batteries");
-  }
-
-  if (encountered_failure) {
-    log_.ERR("STM", "Engaging emergency brakes");
-    sm_data.current_state = data::State::kEmergencyBraking;
-    data_.setStateMachineData(sm_data);
-    state_machine_->current_state_ = state_machine_->failure_braking_;
-  }
+  updateModuleData();
+
+  bool emergency = checkEmergency(log, embrakes_data_, nav_data_, batteries_data_, telemetry_data_,
+                                  sensors_data_, motors_data_);
+  if (emergency) { return FailureBraking::getInstance(); }
+
+  bool stopped = checkPodStopped(log, nav_data_);
+  if (stopped) { return Finished::getInstance(); }
+  return nullptr;
 }
 
-// Finished state
+//--------------------------------------------------------------------------------------
+//  Finished
+//--------------------------------------------------------------------------------------
 
-void Finished::transitionCheck()
+Finished Finished::instance_;
+data::State Finished::enum_value_       = data::kFinished;
+char Finished::string_representation_[] = "Finished";
+
+State *Finished::checkTransition(Logger &log)
 {
-  utils::System &sys             = utils::System::getSystem();
-  data::Telemetry telemetry_data = data_.getTelemetryData();
-
-  if (telemetry_data.shutdown_command) {
-    log_.INFO("STM", "Shutdown command received");
-    log_.INFO("STM", "System is shutting down");
-    sys.running_ = false;
-  }
+  // We only need to update telemetry data.
+  telemetry_data_ = data_.getTelemetryData();
+  if (checkShutdownCommand(log, telemetry_data_)) { return Off::getInstance(); }
+  return nullptr;
 }
 
```

Comment:
Do we ever enter kRunComplete or kExiting?

Reply:
> https://github.com/Hyp-ed/hyped-2021/blob/7dc198fca297a73bf23c0aa7a5563d05787899d5/src/data/data.hpp#L157-L170
> 
> Here I think `kRunComplete`, `kExiting` and `kInvalid` states are never set by state machine, but other subsystems rely on these states.

> That is also a good observation. I think this PR works without fixing that, but I will keep in mind that we have to change that before we merge stm into develop.

Reply:
If I recall correctly, runComplete is when the pod has finished braking, but has not yet begun to crawl out of the tube using service propulsion. Exiting is when the pod is actively waiting for the SVPGO/SVPSTOP commands from telemetry. Finished is when the pod has completed the crawl out the tube.

Reply:
Oh i thought this pr was for a merge into develop. In any case, these states are important for motors, but could be simplified! Maybe a single state for post run, stopped, and a single state for post run, crawling. This would be easy to account for in the propulsion code, but im not sure about other subsystems

Reply:
That sounds like we're going to leave them in then. 

We never actually enter them at the moment since crawling is not a thing, but if you guys have implemented that part already, it will be easier in the future to make it work.